src/pykeen/models/unimodal/triple_re.py

```diff
@@ -0,0 +1,113 @@
+from typing import Any, ClassVar, Mapping, Optional
+
+from torch.nn import functional
+from torch.nn.init import uniform_
+
+from ..nbase import ERModel
+from ...constants import DEFAULT_EMBEDDING_HPO_EMBEDDING_DIM_RANGE
+from ...nn.emb import EmbeddingSpecification
+from ...nn.modules import TripleREInteraction
+from ...typing import Hint, Initializer, Normalizer
+
+__all__ = [
+    "PairRE",
+]
+
+
+class TripleRE(ERModel):
+    r"""An implementation of TripleRE from [yu2021]_.
+
+    ---
+    citation:
+        author: Yu
+        year: 2021
+        link: https://vixra.org/abs/2112.0095
+        github: https://github.com/LongYu-360/TripleRE-Add-NodePiece/
+    """
+
+    #: The default strategy for optimizing the model's hyper-parameters
+    hpo_default: ClassVar[Mapping[str, Any]] = dict(
+        embedding_dim=DEFAULT_EMBEDDING_HPO_EMBEDDING_DIM_RANGE,
+        p=dict(type=int, low=1, high=2),
+    )
+
+    #: The default entity normalizer parameters
+    #: The entity representations are normalized to L2 unit length
+    #: cf. https://github.com/LongYu-360/TripleRE-Add-NodePiece/blob/994216dcb1d718318384368dd0135477f852c6a4/TripleRE%2BNodepiece/ogb_wikikg2/model.py#L323-L324
+    default_entity_normalizer_kwargs: ClassVar[Mapping[str, Any]] = dict(
+        p=2,
+        dim=-1,
+    )
+
+    def __init__(
+        self,
+        embedding_dim: int = 128,
+        p: int = 1,
+        power_norm: bool = False,
+        entity_initializer: Hint[Initializer] = uniform_,
+        entity_initializer_kwargs: Optional[Mapping[str, Any]] = None,
+        entity_normalizer: Hint[Normalizer] = functional.normalize,
+        entity_normalizer_kwargs: Optional[Mapping[str, Any]] = None,
+        relation_initializer: Hint[Initializer] = uniform_,
+        relation_initializer_kwargs: Optional[Mapping[str, Any]] = None,
+        **kwargs,
+    ) -> None:
+        r"""Initialize TripleRE via the :class:`pykeen.nn.modules.TripleREInteraction` interaction.
+
+        :param embedding_dim: The entity embedding dimension $d$.
+        :param p: The $l_p$ norm.
+        :param power_norm: Should the power norm be used?
+        :param entity_initializer: Entity initializer function. Defaults to :func:`torch.nn.init.uniform_`
+        :param entity_initializer_kwargs: Keyword arguments to be used when calling the entity initializer
+        :param entity_normalizer: Entity normalizer function. Defaults to :func:`torch.nn.functional.normalize`
+        :param entity_normalizer_kwargs: Keyword arguments to be used when calling the entity normalizer
+        :param relation_initializer: Relation initializer function. Defaults to :func:`torch.nn.init.uniform_`
+        :param relation_initializer_kwargs: Keyword arguments to be used when calling the relation initializer
+        :param kwargs: Remaining keyword arguments passed through to :class:`pykeen.models.ERModel`.
+        """
+        # TODO: the published code uses NodePiece representations instead
```

Comment:
this requires extension of NodePiece to also support the shortest path / anchor node representations

Reply:
I love how the modularization of PyKEEN allows us to define things like this!

Reply:
do we want to just leave a "TODO" for this, or use a different part of the existing nodepiece implementation instead?

Reply:
We could consider moving the model part into another branch / PR until we have all components needed there. Then, this PR would only have the interaction function :slightly_smiling_face:

Reply:
I'm fine with that

Reply:
-> https://github.com/pykeen/pykeen/pull/715